legion/engine/llri-dx/source.cpp

```diff
@@ -9,88 +10,143 @@ namespace legion::graphics::llri
+            }
+
+            //Attempt to create factory
+            IDXGIFactory* factory = nullptr;
+            const HRESULT factoryCreateResult = CreateDXGIFactory2(factoryFlags, IID_PPV_ARGS(&factory));
+            if (FAILED(factoryCreateResult))
             {
-                destroyInstance(output);
-                return extensionCreateResult;
+                llri::destroyInstance(output);
+                return internal::mapHRESULT(factoryCreateResult);
             }
+
+            //Store factory and return result
+            output->m_ptr = factory;
+            *instance = output;
+            return result::Success;
         }
 
-        //Attempt to create factory
-        IDXGIFactory* factory = nullptr;
-        const HRESULT factoryCreateResult = CreateDXGIFactory2(factoryFlags, IID_PPV_ARGS(&factory));
-        if (FAILED(factoryCreateResult))
+        void impl_destroyInstance(Instance* instance)
         {
-            destroyInstance(output);
-            return internal::mapHRESULT(factoryCreateResult);
-        }
+            if (!instance)
+                return;
 
-        //Store factory and return result
-        output->m_ptr = factory;
-        *instance = output;
-        return result::Success;
-    }
+            for (auto& [ptr, adapter] : instance->m_cachedAdapters)
+                delete adapter;
 
-    void destroyInstance(Instance* instance)
-    {
-        if (!instance)
-            return;
+            if (instance->m_debugAPI)
+                static_cast<ID3D12Debug*>(instance->m_debugAPI)->Release();
 
-        for (auto& [ptr, adapter] : instance->m_cachedAdapters)
-            delete adapter;
+            if (instance->m_debugGPU)
+                static_cast<ID3D12Debug1*>(instance->m_debugGPU)->Release();
 
-        if (instance->m_debugAPI)
-            static_cast<ID3D12Debug*>(instance->m_debugAPI)->Release();
+            delete instance;
+        }
 
-        if (instance->m_debugGPU)
-            static_cast<ID3D12Debug1*>(instance->m_debugGPU)->Release();
+        void impl_pollAPIMessages(const validation_callback_desc& validation, void* messenger)
+        {
+            if (messenger != nullptr && messenger != (void*)1)
+            {
+                ID3D12InfoQueue* iq = static_cast<ID3D12InfoQueue*>(messenger);
+                const auto numMsg = iq->GetNumStoredMessages();
+                
+                for (UINT64 i = 0; i < numMsg; ++i)
+                {
+                    SIZE_T messageLength = 0;
+                    iq->GetMessage(i, NULL, &messageLength);
+                    
+                    D3D12_MESSAGE* pMessage = (D3D12_MESSAGE*)malloc(messageLength);
+                    iq->GetMessage(i, pMessage, &messageLength);
+                    validation(internal::mapSeverity(pMessage->Severity), validation_callback_source::InternalAPI, pMessage->pDescription);
+                    
+                    free(pMessage);
```

Comment:
malloc and free??? no new and delete???

Reply:
This follows the DirectX recommended usage and I'm unsure if it's safe to change to new/delete here (it may very well be). I could try and test with new/delete to see if it sitll functions as expected

Reply:
I don't think this is avoidable because the message size can vary based on the queried message length https://docs.microsoft.com/en-us/windows/win32/api/d3d12sdklayers/nf-d3d12sdklayers-id3d12infoqueue-getmessage

Reply:
Yeah honestly the windows API and dx recommended usage throws in a lot of old convention BS. But tbh i guess since the creation of `pMessage` is handled in `GetMessage`, `malloc` allows you to separate construction and allocation.
normally in c++ you could do raw allocation using:
```cpp
D3D12_MESSAGE* pMessage = reinterpret_cast<D3D12_MESSAGE*>(new std::uint8_t[messageLength]);
```

but depending on what `D3D12_MESSAGE` is, the above might not be necessary, and you might be able to do:
```cpp
D3D12_MESSAGE* pMessage = new D3D12_MESSAGE[messageLength];
```
if for instance `D3D12_MESSAGE` was defined as a `char`, `uchar`, `uint8`, `int8`, or any other byte sized structure without any constructors.

Reply:
D3D12_MESSAGE is an actual struct with data so that last one isn't an option. I think I'll probably stick with malloc/free in this one because in my opinion that's actually more readable than the uint8 array approach. I'll switch the cast from C style to reinterpret to be more explicit

Reply:
fair enough, I can definitely agree with that.